src/Microsoft.AspNetCore.Mvc.Razor/Internal/FileProviderRazorProjectItem.cs

```diff
@@ -27,11 +30,50 @@ public FileProviderRazorProjectItem(IFileInfo fileInfo, string basePath, string
 
         public override string PhysicalPath => FileInfo.PhysicalPath;
 
-        public override string RelativePhysicalPath { get; }
+        public override string RelativePhysicalPath
+        {
+            get
+            {
+                if (_relativePhysicalPath == null)
+                {
+                    if (Exists)
+                    {
+                        if (_root != null && !string.IsNullOrEmpty(PhysicalPath) && PhysicalPath.StartsWith(_root))
+                        {
+                            _relativePhysicalPath = PhysicalPath.Substring(_root.Length + 1); // Include leading separator
+                        }
+                        else
+                        {
+                            // Use FilePath if the file is not directly accessible
+                            _relativePhysicalPath = NormalizeAndEnsureValidPhysicalPath(FilePath);
+                        }
+                    }
+                }
+
+                return _relativePhysicalPath;
+            }
+        }
 
         public override Stream Read()
         {
             return FileInfo.CreateReadStream();
         }
+
+        private static string NormalizeAndEnsureValidPhysicalPath(string filePath)
```

Comment:
Doing this because we want `RelativePhysicalPath` to use OS specific path separator and not have a leading slash.